src/storage-client/src/storage_collections.rs

```diff
@@ -0,0 +1,2816 @@
+    /// be reclaimed.
+    ///
+    /// TODO(jkosh44): This method does not validate the provided identifiers.
+    /// Currently when the controller starts/restarts it has no durable state.
+    /// That means that it has no way of remembering any past commands sent. In
+    /// the future we plan on persisting state for the controller so that it is
+    /// aware of past commands. Therefore this method is for dropping sources
+    /// that we know to have been previously created, but have been forgotten by
+    /// the controller due to a restart. Once command history becomes durable we
+    /// can remove this method and use the normal `drop_sources`.
+    fn drop_collections_unvalidated(
+        &mut self,
+        storage_metadata: &StorageMetadata,
+        identifiers: Vec<GlobalId>,
+    );
+
+    /// Assigns a read policy to specific identifiers.
+    ///
+    /// The policies are assigned in the order presented, and repeated
+    /// identifiers should conclude with the last policy. Changing a policy will
+    /// immediately downgrade the read capability if appropriate, but it will
+    /// not "recover" the read capability if the prior capability is already
+    /// ahead of it.
+    ///
+    /// This [StorageCollections] may include its own overrides on these
+    /// policies.
+    ///
+    /// Identifiers not present in `policies` retain their existing read
+    /// policies.
+    fn set_read_policies(&mut self, policies: Vec<(GlobalId, ReadPolicy<Self::Timestamp>)>);
+
+    /// Acquires and returns the earliest possible read holds for the specified
+    /// collections.
+    fn acquire_read_holds(
+        &mut self,
+        desired_holds: Vec<GlobalId>,
+    ) -> Result<Vec<ReadHold<Self::Timestamp>>, ReadHoldError>;
+
+    /// Applies `updates` and sends any appropriate compaction command.
+    ///
+    /// This is a legacy interface that should _not_ be used! It is only used by
+    /// the compute controller.
+    fn update_read_capabilities(
+        &mut self,
+        updates: &mut BTreeMap<GlobalId, ChangeBatch<Self::Timestamp>>,
+    );
+}
+
+#[derive(Debug, Clone)]
+enum PersistTxns<T> {
+    EnabledEager,
+    EnabledLazy { txns_read: TxnsRead<T> },
+}
+
+impl<T: TimelyTimestamp + Lattice + Codec64> PersistTxns<T> {
+    fn expect_enabled_lazy(&self, txns_id: &ShardId) -> &TxnsRead<T> {
+        match self {
+            PersistTxns::EnabledLazy { txns_read, .. } => {
+                assert_eq!(txns_id, txns_read.txns_id());
+                txns_read
+            }
+            PersistTxns::EnabledEager { .. } => {
+                panic!("set if txns are enabled and lazy")
+            }
+        }
+    }
+}
+
+/// Implementation of [StorageCollections] that is shallow-cloneable and uses a
+/// background task for doing work concurrently, in the background.
+#[derive(Debug, Clone)]
+pub struct StorageCollectionsImpl<
+    T: TimelyTimestamp + Lattice + Codec64 + From<EpochMillis> + TimestampManipulation,
+> {
```

Comment:
I noticed that in contrast to the `StorageController`, for `StorageCollections` the trait impl does not live in a separate crate. AFAIU the purpose of the `StorageController` trait is to free clients from having to depend on the implementation crate. Is that understanding wrong or does the split have a different purpose for `StorageCollections`?

Reply:
That was the motivation, yes! But all of `StorageCollections` is pretty much a "client thing", which is why it's impl can live in the client crate. I did keep the customary separation out of a sense of consistency, plus it _does_ hide away some of the implementation details.

Reply:
Fine for me! I think there is some friction introduced by having this trait in that "go to definition" leads you to the trait method definitions and there is no good way to get from there to the implementation (or at least I haven't found one apart from ctrl+f). But the point about hiding implementation details is also valid.

Reply:
In my setup, I can hover over the method and do a "go to implementation", which takes me to the implementation right away. Maybe you have that as well?

Reply:
Turns out I do! Thanks :D